contracts/Router.sol

```diff
@@ -170,62 +231,71 @@ contract Router is ReentrancyGuard {
 
                 delete(amountsPledged[pledger]);
             }
             delete(pledgers[i]);
         }
 
+	    tree.transferFrom(msg.sender, address(this), amountIn);
+
+	    if (totalInBurnPool > 0) {
+            uint reserveToDistribute = (reserveToken.balanceOf(address(this)).sub(totalPledged)).mul(
+                Babylonian.sqrt(totalInBurnPool)).div(Babylonian.sqrt(treeSupply));
+
+            // Update reserve token claimable for each burner
+            for (uint i = 1; i <= numBurners; i++) {
+                address burner = burners[i];
+                uint256 amountBurned = amountsBurned[burner];
+
+                if (amountBurned > 0) {
+                    // treeToReceive = value pledged * (amountIn / totalPledged)
+                    // For example, if 100 DAI is pledged and there's only 50 TREE available
+                    // an address that pledged 5 DAI would receive 5 * (50/100) = 2.5 TREE
+                    uint256 reserveToReceive = amountBurned.mul(reserveToDistribute).div(totalInBurnPool);
+                    reserveClaimable[burner] = reserveClaimable[burner].add(reserveToReceive);
+
+                    delete (amountsBurned[burner]);
+                }
+                delete (burners[i]);
+            }
+
+		    // Burn the TREE
+		    tree.transfer(address(0), totalInBurnPool);
+        }
+
         // Increase our internal measure of treeSupply by the amount sent in plus the amount sent to LP rewards
-        treeSupply = treeSupply.add(amountIn.div(PRECISION.sub(rewardsCut)));
+        // minus the amount burned
+        treeSupply = treeSupply.add(amountIn.div(PRECISION.sub(rewardsCut))).sub(totalInBurnPool);
 
         if (!hasTransferredOldReserveBalance) {
             // move oldReserveBalance to charity by reversing the code that computes the charityCutAmount
-            // https://github.com/WhalerDAO/tree-contracts/blob/master/contracts/TREEReserve.sol#L173-L175
             amounts[0] = 0;
             amounts[1] = oldReserveBalance.div(charityCut).mul(PRECISION.sub(rewardsCut));
             hasTransferredOldReserveBalance = true;
         } else {
             // send some of the reserveToken to charity
-            uint256 charityCutAmount = reserveTokenReceived.mul(charityCut).div(
-                PRECISION.sub(rewardsCut)
-            );
+	        uint256 charityCutAmount = totalPledged.mul(charityCut).div(PRECISION.sub(rewardsCut));
             reserveToken.safeIncreaseAllowance(address(omniBridge), charityCutAmount);
             omniBridge.relayTokens(address(reserveToken), charity, charityCutAmount);
         }
 
-        emit Rebase(totalPledged, numPledgers);
+        emit Rebase(totalPledged, numPledgers, totalInBurnPool, numBurners);
 
-        // Reset tracking variables
+        // Reset pledging and burning pools
         totalPledged = 0;
         numPledgers = 0;
-    }
-
-
-    function burnTREE(uint256 amount) external nonReentrant {
```

Comment:
Should we still allow `burnTREE()` for solo burners? Maybe someone would want to burn before the rebase

Reply:
The current scheme only works if pooling is the only option.